tests/codecs/test_decorator.py

```diff
@@ -0,0 +1,279 @@
+                ),
+                ResponseOutput(
+                    name="output-1",
+                    datatype="BYTES",
+                    shape=[1],
+                    data=[b"foo"],
+                    parameters=Parameters(content_type=StringCodec.ContentType),
+                ),
+            ],
+        ),
+        (
+            (["foo"], np.array([2])),
+            [StringCodec, NumpyCodec],
+            [
+                ResponseOutput(
+                    name="output-0",
+                    datatype="BYTES",
+                    shape=[1],
+                    data=[b"foo"],
+                    parameters=Parameters(content_type=StringCodec.ContentType),
+                ),
+                ResponseOutput(
+                    name="output-1",
+                    datatype="INT64",
+                    shape=[1],
+                    data=[2],
+                ),
+            ],
+        ),
+        (
+            pd.DataFrame({"a": [2], "b": ["foo"]}),
+            [PandasCodec],
+            [
+                ResponseOutput(name="a", datatype="INT64", shape=[1], data=[2]),
+                ResponseOutput(name="b", datatype="BYTES", shape=[1], data=[b"foo"]),
+            ],
+        ),
+        (
+            (pd.DataFrame({"a": [2], "b": ["foo"]}), ["bar"]),
+            [PandasCodec, StringCodec],
+            [
+                ResponseOutput(name="a", datatype="INT64", shape=[1], data=[2]),
+                ResponseOutput(name="b", datatype="BYTES", shape=[1], data=[b"foo"]),
+                ResponseOutput(
+                    name="output-1",
+                    datatype="BYTES",
+                    shape=[1],
+                    data=[b"bar"],
+                    parameters=Parameters(content_type=StringCodec.ContentType),
+                ),
+            ],
+        ),
+        (
+            (np.array([3]), pd.DataFrame({"a": [2], "b": ["foo"]})),
+            [NumpyCodec, PandasCodec],
+            [
+                ResponseOutput(
+                    name="output-0",
+                    datatype="INT64",
+                    shape=[1],
+                    data=[3],
+                ),
+                ResponseOutput(name="a", datatype="INT64", shape=[1], data=[2]),
+                ResponseOutput(name="b", datatype="BYTES", shape=[1], data=[b"foo"]),
+            ],
+        ),
+    ],
+)
+def test_encode_response(
+    signature_codec: SignatureCodec,
+    output_values: np.ndarray,
+    output_codecs: List[InputCodec],
+    expected_outputs: List[ResponseOutput],
+):
+    signature_codec._output_codecs = output_codecs
+    response = signature_codec.encode_response(model_name="foo", payload=output_values)
+
+    assert response.model_name == "foo"
+    assert response.outputs == expected_outputs
+
+
+@pytest.mark.parametrize(
+    "invalid_values",
+    [
+        ("foo", np.array([2])),
+        (np.array([2]), "foo"),
+    ],
+)
+def test_encode_response_not_found(
```

Comment:
Perhaps useful to add a docstring which part of the logic is being tested as I had to hunt it down in the code - I assume it's because the `signature_codec` fixture has only one output codec corresponding to a `numpy` type but we try to make it encode 2 values, one of which is missing a codec?

Reply:
Good point! Let me add that :+1: